sky/utils/vpn_utils.py

```diff
@@ -0,0 +1,309 @@
+                return device_info.get('id')
+        return None
+
+    def get_private_ip(self, hostname: str) -> str:
+        """Get the private IP address from the hostname."""
+        if self._enable_api:
+            device_id = self._get_device_id_from_hostname(hostname)
+            url_to_query = f'https://api.tailscale.com/api/v2/device/{device_id}'
+            resp = requests.get(url_to_query, headers=self._get_auth_headers())
+            return resp.json().get('addresses', [])[0]
+        # If the API is not enabled, we need to query the IP address locally.
+        # However, since currently self._enable_api is always True, this code
+        # will not be executed.
+        query_cmd = f'tailscale ip -4 {hostname}'
+        rc, stdout, stderr = subprocess_utils.run_with_retries(
+            query_cmd,
+            max_retry=1000,
+            retry_stderrs=['no such host', 'server misbehaving'])
+        subprocess_utils.handle_returncode(
+            rc,
+            query_cmd,
+            error_msg=('Failed to query private IP address'
+                       f' of hostname {hostname}.'),
+            stderr=stdout + stderr)
+        return stdout.strip()
+
+    def remove_host(self, hostname: str) -> None:
+        """Remove a host from the VPN."""
+        if not self._enable_api:
+            logger.warning('API is not enabled. Skipping host removal.')
+            return
+
+        device_id = self._get_device_id_from_hostname(hostname)
+        if not device_id:
+            logger.warning(f'Could not find node ID for hostname {hostname}.'
+                           ' Skipping host removal.')
+
+        url_to_remove = f'https://api.tailscale.com/api/v2/device/{device_id}'
+        resp = requests.delete(url_to_remove, headers=self._get_auth_headers())
+        if resp.status_code != 200:
+            logger.warning(f'Failed to remove host {hostname} from the VPN.'
+                           f' Status code: {resp.status_code}.'
+                           f' Response: {resp.text}')
+
+    @staticmethod
+    def from_backend_config(config: Dict[str, Any]) -> 'TailscaleVPN':
+        assert config.get('auth_key') is not None, (
+            'Tailscale auth key is not set in the configuration.')
+        return TailscaleVPN(**config)
+
+    def to_backend_config(self) -> Dict[str, Any]:
+        return {
+            'auth_key': self._auth_key,
+            'api_key': self._api_key,
+            'network_name': self._network_name,
+            'enable_api': self._enable_api,
+        }
+
+
+class VPNConfig:
+    """VPN Configuration.
+
+    This class is used to manage multiple VPN configurations. It provides
+    a unified interface to connect different VPN configurations and the
+    core logic of SkyPilot. All VPN configurations will be stored in this
+    class, and the core logic will only interact with this class to get
+    installation commands, private IP addresses, etc.
```

Comment:
Do we really need to enable multiple VPN in one cluster? If we have 2 vpn in one cluster, which ip should i use as the public ip? I think we can keep multiple choices of VPN for different cluster and each cluster choose one of them.